devito/core/gpu_openmp.py

```diff
@@ -185,13 +193,28 @@ def place_ondevice(self, iet, **kwargs):
                     writes.add(i.write)
                 reads = (reads | {r for r in i.reads if r.is_DiscreteFunction}) - writes
 
-        # Update `storage`
-        for i in filter_sorted(writes):
-            self._map_function_on_high_bw_mem(i, storage)
-        for i in filter_sorted(reads):
-            self._map_function_on_high_bw_mem(i, storage, read_only=True)
+            # Populate `storage`
+            storage = Storage()
+            for i in filter_sorted(writes):
+                self._map_function_on_high_bw_mem(i, storage)
+            for i in filter_sorted(reads):
+                self._map_function_on_high_bw_mem(i, storage, read_only=True)
```

Comment:
so what happens if something is a write and a read?

Reply:
underneath sets are used and only one declaration is added. This happens in pretty much all tests (`u[t+1, x] = u[t, x]` -- `u` is both a write and a read)